### PR TITLE
Add signout and 'save and sign out' buttons

### PR DESCRIPTION
**Pull request by insacuri**
### What is the context of this PR?
https://trello.com/c/39EwmSSe/2385-sign-out-of-eq-and-ras

Adds a 'sign out' or 'save and sign out' button to blue banner and removed it from the submit button on the page.

### How to review 
Open a survey and sure that for the introduction page and thank you page, the button in the blue banner reads 'Sign out' and every other page should have 'Save and Sign out' as the button in the blue banner.

Additionally, make the browser smaller to trigger the mobile version of the site and it should appear in the dropdown menu also.
### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

**Comment by insacuri** [comment]
Adding in progress label because I forgot to run the tests and a bunch of the broke

File: app/templates/partials/header.html [hunk 1]
```diff
@@ -24,6 +24,13 @@
                 <div class="grid__col {{ 'col-10@xs col-8@s' if account_service_url }}">
                     <h1 class="header__title ">{{ survey_title }}</h1>
                 </div>
+                {% if account_service_log_out_url %}
+                    {% if not content or content.block and content.block.type == 'Introduction' %}
+                    <div class="u-fr u-d-no--@xs@s"><button class="btn btn--ghost print__hidden" data-qa="btn-save-sign-out" type="submit" name="action[save_sign_out]">{{ _("Sign out") }}</button></div>
```
Comment (MebinAbraham) [inline app/templates/partials/header.html]: `data-qa` and `action` needs updating?. Also worth adding event for analytics?.

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
Layout issue:
screenshot 2018-11-05 at 12 37 05

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
Something odd is happening, when you hover over an answer and hit the `Enter` button, it invokes the save and signout button?
Seems to be doing it for like every answer type.


screenshot 2018-11-05 at 12 40 54


Form Data:
```
action[save_sign_out]: 
csrf_token: IjBiNTY1NzVhMjE3MjQ3ZWRmMmM5OGM1NTRjOTI1YjdmZGRhNjhiYWMi.DsHNkw.6Vcxidf0ErDOsTjphli1RduIInU
weekly-pay-paid-employees-answer: 9
```

**Review by MebinAbraham — COMMENTED** [review]
Does this not need to apply for our other themes?

File: app/templates/partials/header.html [hunk 2]
```diff
@@ -24,6 +24,13 @@
                 <div class="grid__col {{ 'col-10@xs col-8@s' if account_service_url }}">
                     <h1 class="header__title ">{{ survey_title }}</h1>
                 </div>
+                {% if account_service_log_out_url %}
+                    {% if not content or content.block and content.block.type == 'Introduction' %}
```
Comment (MebinAbraham) [inline app/templates/partials/header.html]: I found this kinda confusing, can we make use parentheses?.
`{% if not content or (content.block and content.block.type == 'Introduction') %}`

**Review by MebinAbraham — COMMENTED** [review]
The layout column seems to slightly off?
screenshot 2018-11-05 at 13 52 25
According to pattern lib, it should be like this?
screenshot 2018-11-05 at 13 54 32

File: app/templates/partials/header.html (hunk 1 above)
Comment (insacuri) [reply app/templates/partials/header.html]: Agreed, I did all 3 bits in the end

File: app/templates/partials/header.html (hunk 2 above)
Comment (insacuri) [reply app/templates/partials/header.html]: Yea agreed, I added brackets

**Comment by insacuri** [comment]
@MebinAbraham I've fixed the column layout (didn't realise the header had a few new classes against it).
I also sorted out the alignment issue too.
I've spoken to Matt about pressing enter logging you out and he's going to look into it.

**Comment by MebinAbraham** [comment]
My last comment is still an issue when the width is between 500-575px

**Review by MebinAbraham — COMMENTED** [review]
I get a 500 error when I click `Sign out` after submission?
```
File "/Users/abrahm/.virtualenvs/eq-survey-runner-TQNVnq61/lib/python3.6/site-packages/werkzeug/routing.py", line 1580, in match
    raise MethodNotAllowed(valid_methods=list(have_match_for))
werkzeug.exceptions.MethodNotAllowed: 405 Method Not Allowed: The method is not allowed for the requested URL.
127.0.0.1 - - [06/Nov/2018 10:42:30] "POST /questionnaire/test/view_submitted_response/thank-you HTTP/1.1" 500 -
```

**Comment by dwyeradam** [comment]
Lewis has confirmed now 👍 so ignore below.

@insacuri can you please chase up RAS/RH regarding https://github.com/ONSdigital/ons-schema-definitions/pull/39. Until we confirm that they are going to pass the `account_service_log_out_url` we shouldn't really merge this. That said i'll still review this on the assumption that the above PR is fine.

**Comment by insacuri** [comment]
@dwyeradam done, Lewis said it's all good.  Is that enough or do we need another from that team looking at it?

**Review by dwyeradam — DISMISSED** [review]
At the moment if `account_service_log_out_url` isn't passed we don't have a save and sign out button. Shouldn't we still have the button but if the `account_service_log_out_url` isn't available we go to the current signed-out.html page? It seems to do this for social. Otherwise we're stopping respondents from saving and signing out part way through.

**Comment by insacuri** [comment]
@dwyeradam This was a tricky one.  For business surveys we 'should' always have it.  This field is a weird one as it's optional but expected for certain surveys.  Is the correct thing here to have a better fallback option? or can we make that field mandatory for surveys we **know** will have an account service driving the whole thing?

**Comment by insacuri** [comment]
@dwyeradam Removed the conditional 'save and sign out' and it works fine.  Should be good to review now

File: tests/functional/base_pages/theme/northernireland/introduction-page.page.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// >>> WARNING THIS PAGE WAS AUTO-GENERATED - DO NOT EDIT!!! <<<
+const QuestionPage = require('../../question.page');
```
Comment (bitdivision) [inline tests/functional/base_pages/theme/northernireland/introduction-page.page.js]: Don't think these pages should be here?

**Review by bitdivision — DISMISSED** [review]

File: tests/functional/base_pages/theme/northernireland/introduction-page.page.js (hunk 3 above)
Comment (insacuri) [reply tests/functional/base_pages/theme/northernireland/introduction-page.page.js]: Removed them now

**Comment by MebinAbraham** [comment]
Any update from Matt?. Sign out on `Enter` still occurs.

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
If I go to the feedback page (`/feedback`) then click 'Save and sign out`, instead of signing out, it actually submits the form and displays:
```
Submission successful
Thank you for your feedback.

It is now safe to close this window.
```

---

Also if you then click `Sign out`, you get a 500 error.
```
werkzeug.exceptions.MethodNotAllowed: 405 Method Not Allowed: The method is not allowed for the requested URL.
127.0.0.1 - - [12/Nov/2018 12:50:55] "POST /feedback/thank-you HTTP/1.1" 500 -
```

**Comment by ajmaddaford** [comment]
@MebinAbraham @dwyeradam @bitdivision this is ready for re-review.

The changes I've made are:
- Introduced a new top level variable to the templates - `save_on_signout` - this is used to determine if the page is saveable before signing out.
- Refactored the sign out button logic into its own template and added macros for the buttons.
- The CSRF tokens are injected directly in the template rather than via the view context. There was a lot of duplication here previously.
- Added additional functional tests to cover the sign out buttons available on the feedback screen and view submission.

I've raised some issues for further refactoring that I felt would have bloated this pull request - https://github.com/ONSdigital/eq-survey-runner/issues/1918 and https://github.com/ONSdigital/eq-survey-runner/issues/1919.

File: tests/functional/helpers.js [hunk 4]
```diff
@@ -20,8 +22,8 @@ const startCensusQuestionnaire = (schema, sexualIdentity = false, region = 'GB-E
   });
 };
 
-function openQuestionnaire(schema, userId = getRandomString(10), collectionId = getRandomString(10), periodId = '201605', periodStr = 'May 2016', region = 'GB-ENG', language = 'en') {
-  return generateToken(schema, userId, collectionId, periodId, periodStr, region, language)
+function openQuestionnaire(schema, { userId = getRandomString(10), collectionId = getRandomString(10), periodId = '201605', periodStr = 'May 2016', region = 'GB-ENG', language = 'en', sexualIdentity = false, includeLogoutUrl = false } = {}) {
```
Comment (bitdivision) [inline tests/functional/helpers.js:25]: :+1:

File: data/en/test_theme_northernireland.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
+{
+    "mime_type": "application/json/ons/eq",
+    "schema_version": "0.0.1",
+    "data_version": "0.0.2",
+    "survey_id": "0",
+    "title": "Test Northern Ireland Survey",
+    "theme": "social",
```
Comment (MebinAbraham) [inline data/en/test_theme_northernireland.json]: This should be `northernireland`

File: data/en/test_theme_northernireland.json [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                    "title": "What is your favourite breakfast food?",
+                    "type": "General"
+                }],
+                "title": "Favourite breakfast food"
+            }, {
+                "type": "Summary",
+                "id": "summary"
+            }],
+            "id": "group",
+            "title": "Norther Ireland Theme Test"
```
Comment (MebinAbraham) [inline data/en/test_theme_northernireland.json]: An `n` was lost in transition...

**Review by MebinAbraham — CHANGES_REQUESTED** [review]

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
`Sign out` button on the`/feedback` page is submitting the form.

**Review by MebinAbraham — COMMENTED** [review]
@ajmaddaford You should also update the message.pot

File: app/templates/layouts/_base.html [hunk 7]
```diff
@@ -79,28 +80,31 @@
         <div class="skip">
           <a class="skip__link" href="#main">{{ _("Skip to content") }}</a>
         </div>
+        <form class="form qa-questionnaire-form" role="form" method="POST" autocomplete="off" novalidate>
+          <input id="csrf_token" name="csrf_token" type="hidden" value="{{ csrf_token() }}"/>
+          <button class="u-vh" tabindex="-1" type="submit" name="action[save_continue]">Save and continue</button>
```
Comment (MebinAbraham) [inline app/templates/layouts/_base.html]: Needs to be marked for translation.

File: app/themes/census/templates/introduction.html [hunk 8]
```diff
@@ -24,10 +24,7 @@ <h4 class="venus"> Enter your 20 digit unique access code </h4>
     <input id="access-code" class="input input--text " value="" data-qa="input-text" name="access-code">
     <a href="http://www.ons.gov.uk/census/help" target="_blank" >{{_('Help finding your access code')}}</a>
   </div>
-  <form action="" class="form" role="form" method="POST" autocomplete="off" novalidate>
-    {{ content.csrf_token }}
-    <button class="btn qa-btn-get-started" type="submit" name="action[start_questionnaire]">Start survey</button>
-  </form>
+  <button class="btn qa-btn-get-started" type="submit" name="action[start_questionnaire]">Start survey</button>
```
Comment (MebinAbraham) [inline app/themes/census/templates/introduction.html:27]: Needs marking for translation.

File: app/themes/census/templates/introduction.html [hunk 9]
```diff
@@ -24,10 +24,7 @@ <h4 class="venus"> Enter your 20 digit unique access code </h4>
...
  <!-- <div class="lock u-mb-s u-mb-m@s"> -->
     <div class="venus lock__text"><p>Important: Your response is legally required</p></div>
```
Comment (MebinAbraham) [inline app/themes/census/templates/introduction.html]: This was never marked for translation?

File: app/themes/census/templates/introduction.html (hunk 8 above)
Comment (MebinAbraham) [reply app/themes/census/templates/introduction.html:27]: ah its census...

File: tests/functional/spec/checkbox.spec.js [hunk 10]
```diff
@@ -93,7 +93,7 @@ describe('Checkbox with "other" option', function() {
 
   });
 
-  it('Given I have previously added text in other texfiled and saved, when I uncheck other options and select a different checkbox as answer, then the text entered in other field must be wiped.', function() {
+  it('Given I have previously added text in other textfield and saved, when I uncheck other options and select a different checkbox as answer, then the text entered in other field must be wiped.', function() {
```
Comment (MebinAbraham) [inline tests/functional/spec/checkbox.spec.js:96]: 👍

**Review by MebinAbraham — COMMENTED** [review]
Code changes looks good, just that translation comment to fix. Will approve tomorrow morning after a final check.

File: app/templates/signed-out.html [hunk 11]
```diff
@@ -4,12 +4,11 @@
...
 {% block main %}
 
-
-<div class="u-mb-s">
+ <div class="u-mb-s">
     <h1 class="saturn">{{ _("Your survey answers have been saved. You are now signed out") }}.</h1>
 </div>
 
-{% if account_service_url %}
+ {% if account_service_url %}
```
Comment (MebinAbraham) [inline app/templates/signed-out.html]: &nbsp; Dodgy &nbsp; &nbsp; indentation.

File: app/views/session.py [hunk 12]
```diff
@@ -72,6 +72,9 @@ def login():
     if 'account_service_url' in claims and claims.get('account_service_url'):
         cookie_session['account_service_url'] = claims.get('account_service_url')
 
+    if 'account_service_log_out_url' in claims and claims.get('account_service_log_out_url'):
+        cookie_session['account_service_log_out_url'] = claims.get('account_service_log_out_url')
```
Comment (MebinAbraham) [inline app/views/session.py:76]: I know the existing is done like this but can it not just be:
```
 if claims.get('account_service_log_out_url'):
        cookie_session['account_service_log_out_url'] = claims.get('account_service_log_out_url')

File: tests/functional/spec/save_sign_out.spec.js [hunk 13]
```diff
@@ -37,4 +52,108 @@ describe('SaveSignOut', function() {
...
+  it('Given I load the feedback page and an account logout url is not set, then there should be no sign out button', function() {
+    return helpers.openQuestionnaire('test_numbers.json', { includeLogoutUrl: false }).then(() => {
+        return browser
+          .url(FeedbackForm.url())
+          .getUrl().should.eventually.contain('feedback')
+          .isExisting(FeedbackForm.signOut()).should.eventually.be.false;
+      });
+  });
+
+  it('@watch  Given I load the feedback page and I sign out, then I should be on the account service logged out page', function() {
```
Comment (MebinAbraham) [inline tests/functional/spec/save_sign_out.spec.js]: `@watch`

File: tests/functional/spec/save_sign_out.spec.js [hunk 14]
```diff
@@ -3,27 +3,42 @@ const SetMinMax = require('../generated_pages/numbers/set-min-max-block.page.js'
...
+          .getUrl().should.eventually.contain('localhost')
           .back()
           .getSource().should.eventually.contain('Your session has expired');
       });
   });
 
-  it('Given i am returning to the questionnaire, then I am returned to the page I was on and can then complete the survey', function() {
-    return helpers.openQuestionnaire('test_numbers.json', 'test_user', collectionId)
-      .then(() => {
+  it('Given I am returning to the questionnaire, then I am returned to the page I was on and can then complete the survey', function() {
```
Comment (MebinAbraham) [inline tests/functional/spec/save_sign_out.spec.js]: Subjective I guess, but we should be doing `Given, When, Then`?

Comment (ajmaddaford) [reply tests/functional/spec/save_sign_out.spec.js]: Fair point, I've updated all scenarios to be Given, When, Then

**Review by MebinAbraham — APPROVED** [review]
Nice 👍

**Review by bitdivision — COMMENTED** [review]
I've looked through all the code :+1:

I've done some testing, but will finish testing and have a last look on Monday morning.

**Comment by bitdivision** [comment]
@ajmaddaford 	Not a big deal, but:

how frequently are employees paid - monthly wages and salaries survey

The save and signout button is not on the same row as the title.

Ah, @MebinAbraham has actually already raised this.

**Review by bitdivision — APPROVED** [review]

**Review by MebinAbraham — COMMENTED** [review]
@ajmaddaford To fix the header issue you just need to edit:
https://github.com/ONSdigital/eq-survey-runner/blob/1bbb7ea0fef9361d0e0982f9f752a10770e2c489/app/themes/northernireland/templates/partials/header.html#L21
https://github.com/ONSdigital/eq-survey-runner/blob/1bbb7ea0fef9361d0e0982f9f752a10770e2c489/app/templates/partials/header.html#L24

Class is missing: `col-6@s`
So should be: 
``` HTML
<div class="grid__col {{ 'col-10@xs col-6@s col-8@m' if account_service_url }}">
```

**Pull request merged**